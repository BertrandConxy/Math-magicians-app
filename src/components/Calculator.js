import React, { useState } from 'react';
import calculate from './logic/calculate';

const Calculator = () => {
  const [state, setState] = useState({
    total: '0',
    next: '',
    operation: '',
  });

  const clickHandler = (e) => {
    const buttonName = e.target.id;
    const { total, next, operation } = calculate(state, buttonName);
    if (next === null && total === null) {
      setState({ next, total: '0', operation });
    } else {
      setState({ next, total, operation });
    }
  };

  return (
    <div className="Calc-container">
      <div className="display-result" data-testid="display">{state.next || state.total || '0'}</div>
      <button data-testid="AC" type="button" id="AC" className="clear pad" onClick={clickHandler}>AC</button>
      <button type="button" id="+/-" className="sign pad" onClick={clickHandler}>+/-</button>
      <button type="button" id="%" className="modular pad" onClick={clickHandler}>%</button>
      <button type="button" id="รท" className="div operator" onClick={clickHandler}>รท</button>
      <button data-testid="7" type="button" id="7" className="seven pad" onClick={clickHandler}>7</button>
      <button data-testid="8" type="button" id="8" className="eight pad" onClick={clickHandler}>8</button>
      <button type="button" id="9" className="nine pad" onClick={clickHandler}>9</button>
      <button type="button" id="x" className="mult operator" onClick={clickHandler}>x</button>
      <button type="button" id="4" className="four pad" onClick={clickHandler}>4</button>
      <button type="button" id="5" className="five pad" onClick={clickHandler}>5</button>
      <button type="button" id="6" className="six pad" onClick={clickHandler}>6</button>
      <button type="button" id="-" className="sub operator" onClick={clickHandler}>-</button>
      <button type="button" id="1" className="one pad" onClick={clickHandler}>1</button>
      <button type="button" id="2" className="two pad" onClick={clickHandler}>2</button>
      <button type="button" id="3" className="three pad" onClick={clickHandler}>3</button>
      <button data-testid="+" type="button" id="+" className="plus operator" onClick={clickHandler}>+</button>
      <button type="button" id="0" className="zero pad" onClick={clickHandler}>0</button>
      <button type="button" id="." className="period pad" onClick={clickHandler}>.</button>
      <button data-testid="=" type="button" id="=" className="equal operator" onClick={clickHandler}>=</button>
    </div>
  );
};

export default Calculator;
