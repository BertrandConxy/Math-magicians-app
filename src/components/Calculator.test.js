import React from 'react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import Calculator from './Calculator';

describe('calculator rendered properly', () => {
  it('renders properly', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
describe('Calculator responds to user interactions', () => {
  it('Displays the result', () => {
    render(<Calculator />);
    const display = screen.getByTestId('display');
    userEvent.click(screen.getByTestId('7'));
    expect(display.innerHTML).toBe('7');
  });
  it('does the calculation and displays the result', () => {
    render(<Calculator />);
    const display = screen.getByTestId('display');
    userEvent.click(screen.getByTestId('AC'));
    userEvent.click(screen.getByTestId('7'));
    expect(display.innerHTML).toBe('7');
    userEvent.click(screen.getByTestId('+'));
    userEvent.click(screen.getByTestId('8'));
    userEvent.click(screen.getByTestId('='));
    expect(display.innerHTML).toBe('15');
  });
});
