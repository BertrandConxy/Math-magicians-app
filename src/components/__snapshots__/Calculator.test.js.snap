// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator rendered properly renders properly 1`] = `
<div
  className="Calc-container"
>
  <div
    className="display-result"
    data-testid="display"
  >
    0
  </div>
  <button
    className="clear pad"
    data-testid="AC"
    id="AC"
    onClick={[Function]}
    type="button"
  >
    AC
  </button>
  <button
    className="sign pad"
    id="+/-"
    onClick={[Function]}
    type="button"
  >
    +/-
  </button>
  <button
    className="modular pad"
    id="%"
    onClick={[Function]}
    type="button"
  >
    %
  </button>
  <button
    className="div operator"
    id="รท"
    onClick={[Function]}
    type="button"
  >
    รท
  </button>
  <button
    className="seven pad"
    data-testid="7"
    id="7"
    onClick={[Function]}
    type="button"
  >
    7
  </button>
  <button
    className="eight pad"
    data-testid="8"
    id="8"
    onClick={[Function]}
    type="button"
  >
    8
  </button>
  <button
    className="nine pad"
    id="9"
    onClick={[Function]}
    type="button"
  >
    9
  </button>
  <button
    className="mult operator"
    id="x"
    onClick={[Function]}
    type="button"
  >
    x
  </button>
  <button
    className="four pad"
    id="4"
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    className="five pad"
    id="5"
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    className="six pad"
    id="6"
    onClick={[Function]}
    type="button"
  >
    6
  </button>
  <button
    className="sub operator"
    id="-"
    onClick={[Function]}
    type="button"
  >
    -
  </button>
  <button
    className="one pad"
    id="1"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className="two pad"
    id="2"
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className="three pad"
    id="3"
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="plus operator"
    data-testid="+"
    id="+"
    onClick={[Function]}
    type="button"
  >
    +
  </button>
  <button
    className="zero pad"
    id="0"
    onClick={[Function]}
    type="button"
  >
    0
  </button>
  <button
    className="period pad"
    id="."
    onClick={[Function]}
    type="button"
  >
    .
  </button>
  <button
    className="equal operator"
    data-testid="="
    id="="
    onClick={[Function]}
    type="button"
  >
    =
  </button>
</div>
`;
